File: rb/irr.rb
module Projection

  def irr(payments)
    if npv(1, payments) > 0; return 1; end
    upper = 1
    lower = 0
    result = 1
    while result.abs > 0.1
      middle = (upper + lower) / 2.0
      result = npv(middle, payments)
      if result < 0
        upper = middle
      elsif result > 0
        lower = middle
      end
    end
    middle
  end

end

File: irr.js
function irr (payments) 
{
        if (npv(1, payments) > 0) return 1; // ensure that irr calc returns a val
        var interval = [0, 1];
        var result = 1;
        // uses the bisection method to find the root
        while (Math.abs(result) > 0.1)
        {
                var middle = (interval[0] + interval[1]) / 2;
                result = npv(middle, payments);
                if(result < 0) interval = [interval[0], middle];
                if(result > 0) interval = [middle, interval[1]];
        }
        return middle;
}

